### PR TITLE
feat: implement Scala 3 Constant and Union support for string-based literals as enums

### PR DESCRIPTION
**Pull request by ThijsBroersen**
This PR implements support for string-based literal types. Single-value or Union-based. 

Example:
```
case class Foo(aOrB: "A" | "B", optA: Option["A"]) derives Schema

```

### CODE REVIEW

File: doc/endpoint/schemas.md [hunk 1]
```diff
@@ -117,6 +117,23 @@ val s: Schema[StringOrInt] = Schema.derivedUnion[StringOrInt]
 If any of the components of the union type is a generic type, any of its validations will be skipped when validating
 the union type, as it's not possible to generate a runtime check for the generic type.
 
+### Derivation for string-based constant union types
```
Comment (adamw) [inline doc/endpoint/schemas.md:120]: thanks for adding the docs, maybe they should go to https://tapir.softwaremill.com/en/latest/endpoint/enumerations.html though?

File: core/src/test/scala-3/sttp/tapir/SchemaMacroScala3Test.scala [hunk 2]
```diff
@@ -88,6 +88,28 @@ class SchemaMacroScala3Test extends AnyFlatSpec with Matchers:
...
+    // then
+    s.name.map(_.show) shouldBe Some("a_or_b")
+
+    s.schemaType should matchPattern { case SchemaType.SString() => }
+    s.validator should matchPattern { case Validator.Enumeration(List("a", "b"), _, _) => }
+  }
+
+  it should "derive schema for a const as a string-based union type" in {
+    // when
+    val s: Schema["a"] = Schema.constStringToEnum
```
Comment (adamw) [inline core/src/test/scala-3/sttp/tapir/SchemaMacroScala3Test.scala:104]: isn't this a special case of the union-type enumeration? (with a single element)

File: core/src/main/scala-3/sttp/tapir/macros/LowPrioSchemaMacrosVersionSpecific.scala [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package sttp.tapir
+package macros
+
+import sttp.tapir.Schema.SName
+
+import scala.compiletime.*
+import scala.compiletime.ops.any.IsConst
+
+trait LowPrioSchemaMacrosVersionSpecific:
```
Comment (adamw) [inline core/src/main/scala-3/sttp/tapir/macros/LowPrioSchemaMacrosVersionSpecific.scala:9]: should probably be `SchemaCompanionMacrosVersionSpecific`, as `SchemaCompanionMacros` is where the other enumeration-derivation macros are? Also I think we use the "...Extensions" suffix for version-specific code.

File: core/src/main/scala-3/sttp/tapir/macros/union_derivation.scala [hunk 4]
```diff
@@ -0,0 +1,58 @@
+package sttp.tapir.macros
+
+import scala.compiletime.*
+import scala.deriving.*
+import scala.quoted.*
+
+@scala.annotation.implicitNotFound("${A} is not a union of ${T}")
+sealed trait IsUnionOf[T, A]
```
Comment (adamw) [inline core/src/main/scala-3/sttp/tapir/macros/union_derivation.scala:8]: can we make all of these `private[tapir]`? the less we expose to the outside world, the better

File: core/src/test/scala-3/sttp/tapir/SchemaMacroScala3Test.scala (hunk 2 above)
Comment (ThijsBroersen) [reply core/src/test/scala-3/sttp/tapir/SchemaMacroScala3Test.scala:104]: What do you mean? I think a const value should be treated as a single-value enum. The nice thing with string-based enums is that they can be merge, e.g.
```
type AorB = "A" | "B"
type C = "C"
type AorBorC = AorB | C
```
Everything becomes structural, saves a lot of enum headaches

Comment (adamw) [reply core/src/test/scala-3/sttp/tapir/SchemaMacroScala3Test.scala:104]: yes, I agree, I'm wondering why it's a different derivation method than with the two-values case

also, would be great to add a test using the types as in your comment :)

File: core/src/main/scala-3/sttp/tapir/macros/LowPrioSchemaMacrosVersionSpecific.scala (hunk 3 above)
Comment (ThijsBroersen) [reply core/src/main/scala-3/sttp/tapir/macros/LowPrioSchemaMacrosVersionSpecific.scala:9]: I don't get the "...Extensions", do you want to call the lowprio trait 'SchemaCompanionMacrosVersionSpecific' or SchemaCompanionMacrosExtensions'?
I sort of applied the same structure zio-json was using, as I created a PR there yesterday.

File: core/src/main/scala-3/sttp/tapir/macros/union_derivation.scala (hunk 4 above)
Comment (ThijsBroersen) [reply core/src/main/scala-3/sttp/tapir/macros/union_derivation.scala:8]: sure, but they are generic and could be useful 🙃

Comment (adamw) [reply core/src/main/scala-3/sttp/tapir/macros/union_derivation.scala:8]: Well it's not the focus of the tapir library. The less code to maintain in `core`, the better :) Nobody is goign to include `tapir-core` in their project to get `IsUnionOf`

File: core/src/main/scala-3/sttp/tapir/macros/LowPrioSchemaMacrosVersionSpecific.scala (hunk 3 above)
Comment (adamw) [reply core/src/main/scala-3/sttp/tapir/macros/LowPrioSchemaMacrosVersionSpecific.scala:9]: Sorry for not being clear, `SchemaCompanionMacrosExtensions` would follow our naming

**Comment by ThijsBroersen** [comment]
@adamw thanks for the feedback, will process it probably tonight.

File: core/src/test/scala-3/sttp/tapir/SchemaMacroScala3Test.scala (hunk 2 above)
Comment (ThijsBroersen) [reply core/src/test/scala-3/sttp/tapir/SchemaMacroScala3Test.scala]: Turns out it could be simplified 🙃

**Comment by ThijsBroersen** [comment]
Added both Codec and Validator derivations. Schema and Codec reuse the Validator now.

**Pull request merged**